### ISSUE TITLE
How to set opacity independantly of visualised scalars?

### ISSUE THREAD

**Issue by truenicoco** [pyvista/pyvista-support#251]
Let's say I would like to represent half of an ellipsoid as transparent, for instance its points which x coordinates is above zero. Here is the closest I could get.

```py
import numpy as np
import pyvista as pv

ellipsoid = pv.ParametricEllipsoid(10, 5, 5)
ellipsoid.point_arrays["color"] = np.linspace(0,10,len(ellipsoid.points))
ellipsoid.point_arrays["opacity"] = np.where(ellipsoid.points[:, 0] > 0, 0.5, 1)

ellipsoid.plot(scalars="color", n_colors=5, opacity="opacity")
```
![image](https://user-images.githubusercontent.com/1313576/94336521-206f1a80-ffe4-11ea-804b-392ebe4e35a0.png)

There are two problems here:

1. Instead of having the opacity set to 0.5 in the half ellipsoid, it is actually set to zero. Instead of having the opacity set to 1 in the other half, it seems to be 0.5. Browsing the source code, I believe this is because the opacity array is in fact normalized before being used. Can I circumvent this?
2. While the `n_colors` argument is not completely ignored (see the color bar), it does not seem to be used on the mesh.

Is there a way to achieve what I want to do? Do I need to use the `rgb` argument of the the `plotter.add_mesh` method or is there something more convenient?

**Comment by akaszynski** [comment]
You're correct on both counts.  Normalize is non-ideal behavior and ``n_colors`` with opacity is broken.  I've pushed a fix for this at:
https://github.com/pyvista/pyvista/pull/906

If you're fine with using the `master` branch, we can hold off a bit before pushing a patch release.  Otherwise, we can make one if you really need the release.

**Comment by truenicoco** [comment]
Wow, thanks for the reactivity on the week-end. :)

I tried your patch and it does indeed fix both problems in my case… but introduced a new one. The half part of the ellipsoid that is supposed to be opaque is slightly transparent now. I also noticed this happening when trying to pass a `n_points×4` array representing RGBA colors as `scalars` and using `rgb=True`. It seems that whenever one part of the mesh is not opaque, no part can be completely opaque. I am wondering if this is a VTK limitation.

**Comment by truenicoco** [comment]
In my real-use case, I am in fact calling `plotter.add_mesh`. You patch has the same bogus behaviour (transparency on parts that are supposed to be opaque), but also crashes when using `smooth_shading=True`, throwing a `RuntimeError: Transfer function cannot have more values than n_colors`.

**Comment by akaszynski** [comment]
> I tried your patch and it does indeed fix both problems in my case… but introduced a new one. The half part of the ellipsoid that is supposed to be opaque is slightly transparent now. I also noticed this happening when trying to pass a `n_points×4` array representing RGBA colors as `scalars` and using `rgb=True`. It seems that whenever one part of the mesh is not opaque, no part can be completely opaque. I am wondering if this is a VTK limitation.

I had to upgrade from ``vtk==8.1.2`` to ``vtk==9.0.1`` to get this behavior.  If possible, download for the time being and see if this fixes the problem and the post an issue on https://gitlab.kitware.com/vtk/vtk

> In my real-use case, I am in fact calling `plotter.add_mesh`. You patch has the same bogus behaviour (transparency on parts that are supposed to be opaque), but also crashes when using `smooth_shading=True`, throwing a `RuntimeError: Transfer function cannot have more values than n_colors`.

This worked on my end (on vtk 9 and 8.1.2):

```py
import numpy as np
import pyvista as pv

ellipsoid = pv.ParametricEllipsoid(10, 5, 5)
ellipsoid.point_arrays["color"] = np.linspace(0,10,len(ellipsoid.points))
ellipsoid.point_arrays["opacity"] = np.where(ellipsoid.points[:, 0] > 0, 0.5, 1)

pl = pv.Plotter()
pl.add_mesh(ellipsoid, scalars="color", n_colors=5, opacity="opacity")
pl.show()
```

**Comment by truenicoco** [comment]
Thanks (again) for your reply.

Opacity-where-it-should-be-opaque was fixed by downgrading VTK to v8.

About the `RuntimeError`, it happened because:

1. I was using a tetrahedral mesh
2. Instead of passing a `str` as opacity, I was using `opacity=mesh.point_arrays['opacity']`. I think this caused the "colors scalars" to be cell data and "opacity scalars" to be point data, though I did not manage to track down what was happening exactly.

Getting back to what I was trying to achieve, I'm still not there. I cannot share the mesh for legal reasons, but I guess a screenshot's OK.

The left part is supposed to be uniformly transparent here:

![image](https://user-images.githubusercontent.com/1313576/94394362-73b8a880-015d-11eb-9b5b-b66122f5956c.png)

It seems that some tetrahedra (or triangles?) remain opaque.

And something worse happen from a different point of view:

![image](https://user-images.githubusercontent.com/1313576/94394480-bbd7cb00-015d-11eb-899a-77775ceb970c.png)

Here the transparent part is between the camera and the opaque part. Transparency, instead of allowing to see the color scalars on the opaque part (which is my motivation for this), seems to make the opaque part weirdly transparent too.

I don't know if depth peeling or other parameters that I don't really understand must be tuned to achieve what I need…

**Comment by akaszynski** [comment]
Seems that enabling depth peeling on VTK>=9.0.1 is your best bet:

```py
import numpy as np
import pyvista as pv

ellipsoid = pv.ParametricEllipsoid(10, 5, 5)
ellipsoid.point_arrays["color"] = np.linspace(0,10,len(ellipsoid.points))
ellipsoid.point_arrays["opacity"] = np.where(ellipsoid.points[:, 0] > 0, 0.5, 1)

pl = pv.Plotter()
pl.enable_depth_peeling()
pl.add_mesh(ellipsoid, scalars="color", n_colors=5, opacity="opacity", smooth_shading=True)
pl.show()
```

Depth peeling doesn't seem to work well on vtk==8.1.2, but on windows with vtk 9.0 it works great.  I've not checked with Linux at the moment, but I'm guessing the behavior is similar.

**Comment by imtamo** [comment]
I would like to do volume rendering of 3d raster data. Raster data is represented by integer labels from 0 to 5. And I would like to adjust opacity and color according to labels. But I cant now. Does anyone have a solution? Thank you.
```python
import pyvista as pv
import numpy as np

grid = pv.ImageData()
grid.dimensions = np.array(data.shape) + 1
grid.origin = [85, 32, 150]
grid.spacing = [0.25 for _ in range(3)]
grid["data"] = data.ravel(order='F')
grid.plot(volume=True, show_grid=True, opacity=0.1, cmap="bone")
```
<img width="769" alt="スクリーンショット 2023-08-24 0 22 14" src="https://github.com/pyvista/pyvista-support/assets/120159576/b373bd6d-db1e-40b6-b665-8da729c29c4b">

**Comment by josephko91** [comment]
@imtamo I would also like to do something similar. Interested in any potential solutions.

EDIT: It looks like there's actually documentation that shows how to do this, with examples. 
https://docs.pyvista.org/version/stable/examples/02-plot/opacity.html
